File: tests/csharp/Hosting.cs
using LanguageExt;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Hosting;
using System;
using System.Diagnostics;
using System.Linq;
using System.Reflection;

namespace common.tests;

#pragma warning disable CA1822 // Mark members as static: Bug https://github.com/dotnet/roslyn-analyzers/issues/7646
internal static class HostingExtensions
{
    extension(IConfiguration configuration)
    {
        public Option<string> GetValue(string key)
        {
            var section = configuration.GetSection(key);

            return section.Exists()
                    ? string.IsNullOrWhiteSpace(section.Value)
                        ? Option<string>.None
                        : section.Value
                    : Option<string>.None;
        }

        public string GetRequiredValue(string key) =>
            configuration.GetValue(key)
                         .IfNone(() => throw new InvalidOperationException($"Configuration key '{key}' not found."));
    }

    extension(IConfigurationBuilder builder)
    {
        public IConfigurationBuilder AddUserSecretsWithLowestPriority(Assembly assembly)
        {
            var newSources =
                new ConfigurationBuilder()
                .AddUserSecrets(assembly, optional: true)
                .Sources
                .ToList();

            var existingSources = builder.Sources.ToList();

            builder.Sources.Clear();

            foreach (var source in newSources.Concat(existingSources))
            {
                builder.Sources.Add(source);
            }

            return builder;
        }
    }
}
#pragma warning restore CA1822 // Mark members as static

File: tests/fsharp/Hosting.fs
namespace common.tests

open System
open System.Diagnostics
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.DependencyInjection.Extensions
open Microsoft.Extensions.Hosting

[<RequireQualifiedAccess>]
module Configuration =
    let getValue key (configuration: IConfiguration) =
        let section = configuration.GetSection(key)

        if section.Exists() then
            Option.ofObj section.Value
        else
            None

    let getRequiredValue key configuration =
        configuration
        |> getValue key
        |> Option.defaultWith (fun () -> failwithf $"Configuration key '{key}' not found.")

    // Adds user secrets to the configuration builder. Added user secrets will have the lowest priority.
    let addUserSecrets assembly (builder: IConfigurationBuilder) =
        let newSources =
            ConfigurationBuilder().AddUserSecrets(assembly, optional = true).Sources
            |> List.ofSeq

        let existingSources = builder.Sources |> List.ofSeq

        builder.Sources.Clear()

        newSources @ existingSources
        |> Seq.iter (fun source -> builder.Sources.Add(source) |> ignore)

        builder

[<RequireQualifiedAccess>]
module HostApplicationBuilder =
    let tryAddSingleton<'a when 'a: not null and 'a: not struct>
        (f: IServiceProvider -> 'a)
        (builder: IHostApplicationBuilder)
        =
        builder.Services.TryAddSingleton<'a>(Func<IServiceProvider, 'a>(f))
        builder

[<RequireQualifiedAccess>]
module ServiceProvider =
    let getService<'a when 'a: not null> (provider: IServiceProvider) = provider.GetRequiredService<'a>()

[<RequireQualifiedAccess>]
module Host =
    let start (host: IHost) =
        async {
            let! cancellationToken = Async.CancellationToken
            do! host.StartAsync(cancellationToken) |> Async.AwaitTask
        }

[<RequireQualifiedAccess>]
module ActivitySource =
    let configureBuilder (name: string) builder =
        builder
        |> HostApplicationBuilder.tryAddSingleton (fun _ -> new ActivitySource(name))
